### ISSUE TITLE
About the test results 

### ISSUE THREAD

**Issue by Dingpx** [wei-mao-2019/LearnTrajDep#5]
@wei-mao-2019 
Hi~ Thanks for sharing the code of this perfect work! 
I just have two questions about the test result (Tables2) in the paper and the code. 
1. Why we just randomly sample four sequences in one txt file instead of evaluate the results in whole S5 datasets?
2. Why the results in val datasets perform better than test datasets?(It seems that the results differ a lot).

**Comment by wei-mao-2019** [comment]
Hi @Dingpx 

#1
We follow the same test protocol as state-of-the-art methods for fair comparison. You can evaluate on the whole sequence.

#2 
Can you be more specific? Did you mean that the error of validation set is much lower than the error of test set in the csv log file?

Thank you
Wei

**Comment by jutanke** [comment]
@Dingpx 
The problems with using only 4 subsequences per action where discussed in this paper (Section 4.2): [QuaterNet](https://arxiv.org/pdf/1901.07677.pdf). Maybe its useful for you..?

@wei-mao-2019 
Could you clarify these lines: 

https://github.com/wei-mao-2019/LearnTrajDep/blob/47530723316f543a3732db611420833aee7054b2/main_3d.py#L223

and

https://github.com/wei-mao-2019/LearnTrajDep/blob/47530723316f543a3732db611420833aee7054b2/main_3d.py#L232-L233

Are the joints that are not predicted by the model taken from the ground-truth?
Is the final 3D loss then calculated over a mix of _predictions + ground truth_ __VS__ _ground truth_ or am I missing something?

**Comment by wei-mao-2019** [comment]
Hi @jutanke,

Thank you for you reply.

One pose consists of 32 joints in H36m dataset and there are two kinds of special joints,

- Joints that never move after eliminate the global translation and rotation. Such as pelvis, left and right hip.

- Joints that have same 3D position all the time. So that only need to predict one 3d position for those have same 3D positions.

Those code is to take GT position for joints that never move and copy predicted 3D positions to those that have the same positions.

Cheers
Wei

**Comment by jutanke** [comment]
Hey Wei,
thanks for the quick response! I am just wondering if it is sensible to calculate the final metric over those joints or if it would be _more fair_ to only calculate it over the 22(?) valid joints that where predicted by the network.
Assuming you compare your method with one that predicts all 32 joints, which has to figure out duplicates etc., it would be at an disadvantage as the evaluation, as it is now, seems to leak ground truth into the prediction.
Cheers
Julian

**Comment by Dingpx** [comment]
@jutanke 

Thanks for your sharing! 
It really helps me a lot.

@wei-mao-2019 

Thanks for your reply! 

I conduct experiments using different models. I use the same evaluation method in val datasets and i find the average loss of 4 short frames(80ms,160ms,320ms,400ms) is lower than test datasets. 

I think there exists the difference because the two datasets represent different people. So it may lead to inherent difference. But i am not sure about it.

**Comment by wei-mao-2019** [comment]
Hi,

@jutanke

For fair comparison, all baseline methods we compared with use same number of joints as us. And yes, you are right. We could eliminate joints that do not need to predict during evaluation Or follow the human pose estimation task using 17 joints. 

@Dingpx 

It is normal that validation set have different performance as test set, but the difference should not be significant. Can you show your evaluation results on validation set and test set?

Thank you
Wei

**Comment by Dingpx** [comment]
Hi,

@wei-mao-2019 

In order to process test dataset the same as train and val datasets i modify "main_3d.py" and corresponding codes. Part of results are as follows.

`>>> epoch: 12 | lr: 0.00039
train_3d:17.67854016536196
val_3d:15.00204092394404
test_3d:18.292376285511622`

So we can find that the average results of 10 frames of val and test have 3 millimeter gap. If a model gets a result like " val_3d : 18.3, test_3d : 15.0 ", how can we judge which model is better? Of course, maybe all models perform better in test dataset. If in this case, why has this happened?

Thank you 
Ding

**Comment by wei-mao-2019** [comment]
Hi @Dingpx,

Since the test set and val set is two different subjects (sub 5 and sub 11). The performance on them can not be exactly the same.

The best model is selected by the error on val set.

Cheers
Wei

**Comment by Dingpx** [comment]
Hi @wei-mao-2019 ,

Thanks for your response. Your guidance helps me a lot.

Cheers 
Ding
